### PR TITLE
feat(install): rewrite posh script & opt-in marketplace install

### PR DESCRIPTION
**Pull request by SunsetTechuila**
test carefully and feel free to suggest changes

### CODE REVIEW

**Comment by afonsojramos** [comment]
I'm very wary of changing our install scripts at this point in time, because they currently work well, and we'll probably move off of install scripts for v3. If I have access to a windows machine in the next week I'll try to test it out!

**Comment by afonsojramos** [comment]
We should probably also update the documentation for the installation then, no? 
And also get the `.sh` installation to also query for Marketplace installation, way better implementation than simply giving two scripts directly for sure.

**Comment by SunsetTechuila** [comment]
> We should probably also update the documentation for the installation then, no?
yes

**Comment by SunsetTechuila** [comment]
> And also get the .sh installation to also query for Marketplace installation,

i don't know bash

**Comment by SunsetTechuila** [comment]
> And also get the .sh installation to also query for Marketplace installation,
@afonsojramos done

**Comment by afonsojramos** [comment]
@SunsetTechuila is this expected behaviour?
image

**Comment by afonsojramos** [comment]
Shell one looks good tho!

**Comment by SunsetTechuila** [comment]
> is this expected behaviour?

no, have fixed it

**Comment by afonsojramos** [comment]
@SunsetTechuila getting this at the marketplace section 
image

**Comment by SunsetTechuila** [comment]
@afonsojramos fixed

image

**Review by afonsojramos — CHANGES_REQUESTED** [review]
@SunsetTechuila The only thing that I don't quite like is that most messages show up as "verbose". That doesn't make that much sense if the "Success" afterwards doesn't... Let's keep them as normal messages, like we've had.

image

PS: This was ran in a windows pc that doesn't have Spotify installed.

**Comment by SunsetTechuila** [comment]
@afonsojramos is it better now?

image

**Review by afonsojramos — APPROVED** [review]
I feel like it is perfect now, thanks @SunsetTechuila! Requesting some other eyes as well

**Review by rxri — CHANGES_REQUESTED** [review]
- The install.log is missing, and it was really useful feature, please bring it back
- Colors are not really nice compared to the older script. Personally, I would switch to the old ones
- It would be nice to highlight the `spicetify-h` and change the message containing it because it says `[...] to get started`, but `-h` just shows help message so it isn't really clear
- Also it seems like marketplace script has problem with detecting if spicetify is already applied?
image

**Comment by SunsetTechuila** [comment]
> The install.log is missing, and it was really useful feature, please bring it back

I've never seen anyone use it

**Comment by rxri** [comment]
> > The install.log is missing, and it was really useful feature, please bring it back
> 
> I've never seen anyone use it

v3 will definitely use it and that's how we will ask people to send install logs. Removing it now, isn't the best idea then

**Comment by SunsetTechuila** [comment]
> v3 will definitely use it 

> [we'll probably move off of install scripts for v3](https://github.com/spicetify/spicetify-cli/pull/2663#issuecomment-1816270321)

image

ummm?

**Comment by rxri** [comment]
Yes, but the installer will generate the log xd

**Comment by SunsetTechuila** [comment]
and how this makes removing logging that no one uses a bad idea?

**Comment by SunsetTechuila** [comment]
image

4 results. support server exists since 2021

**Comment by SunsetTechuila** [comment]
> and change the message containing it

to what exactly

**Comment by SunsetTechuila** [comment]
> Colors are not really nice compared to the older script. Personally, I would switch to the old ones

I used cyan so that the info lines about current actions would not merge with the prompts
Do you really insist on making them white?

File: install.ps1 [hunk 1]
```diff
@@ -1,177 +1,202 @@
...
+  end {
+    $archivePath
+  }
+}
+
+function Add-SpicetifyToPath {
+  [CmdletBinding()]
+  param ()
+  begin {
+    Write-Host -Object 'Adding Spicetify to your PATH if needed...' -NoNewline
```
Comment (rxri) [inline install.ps1:110]:
```suggestion
    Write-Host -Object 'Making spicetify available in the PATH...' -NoNewline

```

File: install.ps1 [hunk 2]
```diff
@@ -1,177 +1,202 @@
...
+#endregion Spicetify
+
+#region Marketplace
+$Host.UI.RawUI.Flushinputbuffer()
+$choice = $Host.UI.PromptForChoice('', "`nDo you want to install Spicetify Marketplace?", ('&Yes', '&No'), 0)
+if ($choice -eq 1) {
+  Write-Host -Object 'Spicetify Marketplace installation aborted' -ForegroundColor 'Yellow'
+}
+else {
+  Write-Host -Object 'Starting the Spicetify Marketplace installation script..'
```
Comment (rxri) [inline install.ps1:194]:
```suggestion
  Write-Host -Object 'Starting the spicetify Marketplace installation script..'

```

File: install.ps1 [hunk 3]
```diff
@@ -1,177 +1,202 @@
...
+Write-Host -Object "`nRun" -NoNewline
+Write-Host -Object ' spicetify -h ' -NoNewline -ForegroundColor 'Cyan'
+Write-Host -Object 'to get started'
+#endregion Spicetify
+
+#region Marketplace
+$Host.UI.RawUI.Flushinputbuffer()
+$choice = $Host.UI.PromptForChoice('', "`nDo you want to install Spicetify Marketplace?", ('&Yes', '&No'), 0)
+if ($choice -eq 1) {
+  Write-Host -Object 'Spicetify Marketplace installation aborted' -ForegroundColor 'Yellow'
```
Comment (rxri) [inline install.ps1:191]:
```suggestion
  Write-Host -Object 'spicetify Marketplace installation aborted' -ForegroundColor 'Yellow'

```

File: install.ps1 [hunk 4]
```diff
@@ -1,177 +1,202 @@
...
+else {
+  Write-Success
+}
+if (-not (Test-Admin)) {
+  Write-Unsuccess
+  Write-Warning -Message "The script was ran as Administrator which isn't recommended"
+  $Host.UI.RawUI.Flushinputbuffer()
+  $choice = $Host.UI.PromptForChoice('', 'Do you want to abort the installation process to avoid any issues?', ('&Yes', '&No'), 0)
+  if ($choice -eq 0) {
+    Write-Host -Object 'Spicetify installation aborted' -ForegroundColor 'Yellow'
```
Comment (rxri) [inline install.ps1:169]:
```suggestion
    Write-Host -Object 'spicetify installation aborted' -ForegroundColor 'Yellow'

```

File: install.ps1 [hunk 5]
```diff
@@ -1,177 +1,202 @@
...
+  process {
+    $archivePath = Get-Spicetify
+    Write-Host -Object 'Extracting Spicetify...' -NoNewline
+    Expand-Archive -Path $archivePath -DestinationPath $spicetifyFolderPath -Force
+    Write-Success
+    Add-SpicetifyToPath
+  }
+  end {
+    Remove-Item -Path $archivePath -Force
+    Write-Host -Object 'Spicetify was successfully installed!' -ForegroundColor 'Green'
```
Comment (rxri) [inline install.ps1:142]:
```suggestion
    Write-Host -Object 'spicetify was successfully installed!' -ForegroundColor 'Green'

```

File: install.ps1 [hunk 6]
```diff
@@ -1,177 +1,202 @@
...
+
+function Install-Spicetify {
+  [CmdletBinding()]
+  param ()
+  begin {
+    Write-Host -Object 'Installing Spicetify...'
+  }
+  process {
+    $archivePath = Get-Spicetify
+    Write-Host -Object 'Extracting Spicetify...' -NoNewline
```
Comment (rxri) [inline install.ps1:135]:
```suggestion
    Write-Host -Object 'Extracting spicetify...' -NoNewline

```

File: install.ps1 [hunk 7]
```diff
@@ -1,177 +1,202 @@
...
+    $env:PATH = $path
+    Write-Success
+  }
+}
+
+function Install-Spicetify {
+  [CmdletBinding()]
+  param ()
+  begin {
+    Write-Host -Object 'Installing Spicetify...'
```
Comment (rxri) [inline install.ps1:131]:
```suggestion
    Write-Host -Object 'Installing spicetify...'

```

File: install.ps1 [hunk 8]
```diff
@@ -1,177 +1,202 @@
...
+    else {
+      Write-Host -Object 'Fetching the latest Spicetify version...' -NoNewline
+      $latestRelease = Invoke-RestMethod -Uri 'https://api.github.com/repos/spicetify/spicetify-cli/releases/latest'
+      $targetVersion = $latestRelease.tag_name -replace 'v', ''
+      Write-Success
+    }
+    $archivePath = "$env:TEMP\spicetify.zip"
+  }
+  process {
+    Write-Host -Object "Downloading Spicetify v$targetVersion..." -NoNewline
```
Comment (rxri) [inline install.ps1:92]:
```suggestion
    Write-Host -Object "Downloading spicetify v$targetVersion..." -NoNewline

```

File: install.ps1 [hunk 9]
```diff
@@ -1,177 +1,202 @@
...
+        $targetVersion = $v
+      }
+      else {
+        Write-Warning -Message "You have spicefied an invalid Spicetify version: $v `nThe version must be in the following format: 1.2.3"
+        Pause
+        exit
+      }
+    }
+    else {
+      Write-Host -Object 'Fetching the latest Spicetify version...' -NoNewline
```
Comment (rxri) [inline install.ps1:84]:
```suggestion
      Write-Host -Object 'Fetching the latest spicetify version...' -NoNewline

```

File: install.ps1 [hunk 10]
```diff
@@ -1,177 +1,202 @@
...
+    }
+    else {
+      $architecture = 'x32'
+    }
+    if ($v) {
+      if ($v -match '^\d+\.\d+\.\d+$') {
+        $targetVersion = $v
+      }
+      else {
+        Write-Warning -Message "You have spicefied an invalid Spicetify version: $v `nThe version must be in the following format: 1.2.3"
```
Comment (rxri) [inline install.ps1:78]:
```suggestion
        Write-Warning -Message "You have spicefied an invalid spicetify version: $v `nThe version must be in the following format: 1.2.3"

```

File: install.ps1 [hunk 11]
```diff
@@ -1,177 +1,202 @@
...
+    $PSVersionTable.PSVersion -ge $PSMinVersion
+  }
+}
+
+function Move-OldSpicetifyFolder {
+  [CmdletBinding()]
+  param ()
+  process {
+    if (Test-Path -Path $spicetifyOldFolderPath) {
+      Write-Host -Object 'Moving your old Spicetify folder...' -NoNewline
```
Comment (rxri) [inline install.ps1:55]:
```suggestion
      Write-Host -Object 'Moving the old spicetify folder...' -NoNewline

```

File: install.sh [hunk 12]
```diff
@@ -135,3 +135,11 @@ esac
 echo
 log "spicetify v$tag was installed successfully to $spicetify_install"
 log "Run 'spicetify --help' to get started"
+
+read -p "Do you want to install Spicetify Marketplace? (Y/n) " choice
+if [[ $choice == [Nn]* ]]; then
+	echo "Spicetify Marketplace installation aborted"
```
Comment (rxri) [inline install.sh:141]:
```suggestion
	echo "spicetify Marketplace installation aborted"
```

File: install.sh [hunk 13]
```diff
@@ -135,3 +135,11 @@ esac
 echo
 log "spicetify v$tag was installed successfully to $spicetify_install"
 log "Run 'spicetify --help' to get started"
+
+read -p "Do you want to install Spicetify Marketplace? (Y/n) " choice
```
Comment (rxri) [inline install.sh:139]:
```suggestion
read -p "Do you want to install spicetify Marketplace? (Y/n) " choice
```

File: install.sh [hunk 14]
```diff
@@ -135,3 +135,11 @@ esac
...
+if [[ $choice == [Nn]* ]]; then
+	echo "Spicetify Marketplace installation aborted"
+	exit
+fi
+echo "Starting the Spicetify Marketplace installation script.."
```
Comment (rxri) [inline install.sh:144]:
```suggestion
echo "Starting the spicetify Marketplace installation script.."
```

**Review by rxri — CHANGES_REQUESTED** [review]

**Review by rxri — APPROVED** [review]

**Pull request merged**